### PR TITLE
Refactored and moved to its own method code responsible for sending VIP entries

### PR DESCRIPTION
**Pull request by elderapo**
Refactored and moved to its own method code responsible for sending VIP entries.

### CODE REVIEW

**Comment by ranisalt** [comment]
Why do we use a forward_list? It is worse than a vector space-wise and slower because iteration is not over contiguous memory. It should be change because the collection is not changed.

File: src/protocolgame.cpp [hunk 1]
```diff
@@ -2703,6 +2675,25 @@ void ProtocolGame::sendVIP(uint32_t guid, const std::string& name, const std::st
 	writeToOutputBuffer(msg);
 }
 
+void ProtocolGame::sendVIPEntries() {
```
Comment (jo3bingham) [inline src/protocolgame.cpp]: Move `{` to new line.

File: src/protocolgame.cpp [hunk 2]
```diff
@@ -2703,6 +2675,25 @@ void ProtocolGame::sendVIP(uint32_t guid, const std::string& name, const std::st
...
+	const std::forward_list<VIPEntry>& vipEntries = IOLoginData::getVIPEntries(player->getAccount());
+
+	for (const VIPEntry& entry : vipEntries) {
+		VipStatus_t vipStatus;
```
Comment (jo3bingham) [inline src/protocolgame.cpp]: May as well assign `VIPSTATUS_ONLINE` to `vipStatus` here, and remove the `else` statement below.

File: src/protocolgame.cpp [hunk 3]
```diff
@@ -2703,6 +2675,25 @@ void ProtocolGame::sendVIP(uint32_t guid, const std::string& name, const std::st
...
+
+		Player* vipPlayer = g_game.getPlayerByGUID(entry.guid);
+
+		if (!vipPlayer || vipPlayer->isInGhostMode() || player->isAccessPlayer()) {
+			vipStatus = VIPSTATUS_OFFLINE;
+		}
+		else {
```
Comment (jo3bingham) [inline src/protocolgame.cpp]: If you prefer to keep this `else` statement (see my comment above), please move it up to the previous line (`} else {`).

**Comment by djarek** [comment]
@ranisalt 
I agree about the forward list, however, I think that should be a separate patch.

**Comment by jo3bingham** [comment]
@ranisalt I'm going to assume that, whomever decided to use forward_list here did so because we only ever traverse forward through this list (which saves space since it only points to the next element and it doesn't contain a size member).

**Comment by ranisalt** [comment]
@jo3bingham I don't think so, a size member is only as big as a pointer, as the function is inlined, so there is absolutely no advantage on having a forward list over a vector *in this case* if the list ever has an element. See:

For 0 elements: forward_list has 1 pointer, vector has 1 pointer and 1 size_t
For 1 element: forward_list has 2 pointers and 1 T, vector has 1 pointer, 1 T and 1 size_t (the same)
For 2 or + elements: forward_list has N+1 pointers and N T, vector has 1 pointer, N T and 1 size_t (vector has N-1 less pointer-sized elements)

Of course, this totally disregard other internal variables, I'm only considering the simplest possible implementation.

**Comment by djarek** [comment]
@ranisalt 
You forgot to mention that general-purpose allocators are pretty bad space-wise in allocating small blocks of memory (in a good allocator you'll have 1 pointer of overhead per allocated block, however, you usually can't allocate less than a fixed size, IIRC in glibc it's 16 bytes). This makes node based containers look really bad both space and performance wise when container traversal is more frequent than insertion/deletion. There are few places were lists *might* be better than contiguous memory storage - when used as queues(pop/push_back/front) or when you do splicing. There are a few places in tfs where forward_list is used to save space when the list itself is heap allocated and represents a container that's usually empty. These should probably be replaced by a "vector" that stores most of its control block along with the data.

File: src/protocolgame.h [hunk 4]
```diff
@@ -225,6 +225,7 @@ class ProtocolGame final : public Protocol
 
 		void sendUpdatedVIPStatus(uint32_t guid, VipStatus_t newStatus);
 		void sendVIP(uint32_t guid, const std::string& name, const std::string& description, uint32_t icon, bool notify, VipStatus_t status);
+		void ProtocolGame::sendVIPEntries();
```
Comment (djarek) [inline src/protocolgame.h]: Remove ProtocolGame:: from this line

**Comment by jo3bingham** [comment]
@ranisalt I was only pointing out a possible reason why someone would think using a forward_list here would be ideal. I agree that the use of a vector would be better.

**Comment by djarek** [comment]
Thanks!

**Pull request merged**